src/search/pdbs/pattern_collection_generator_multiple_cegar.cc

```diff
@@ -0,0 +1,300 @@
+        }
+        if (!is_goal_var) {
+            non_goal_variables.push_back(var_id);
+        }
+    }
+
+    if (verbosity >= utils::Verbosity::DEBUG) {
+        utils::g_log << "goal variables: ";
+        for (auto goal : goals) {
+            utils::g_log << goal.var << ", ";
+        }
+        utils::g_log << endl;
+        utils::g_log << "non-goal variables: " << non_goal_variables << endl;
+    }
+
+    // Collect all unique patterns and their PDBs.
+    utils::HashSet<Pattern> generated_patterns;
+    shared_ptr<PDBCollection> generated_pdbs = make_shared<PDBCollection>();
+
+    bool blacklisting = false;
+    double stagnation_start_time = -1;
+    int num_iterations = 1;
+    int goal_index = 0;
+    int remaining_collection_size = max_collection_size;
+    const utils::Verbosity cegar_verbosity(utils::Verbosity::SILENT);
+    /*
+      Start blacklisting after the percentage of total_max_time specified via
+      blacklisting_trigger_percentage has passed. Compute this time point once.
+    */
+    double blacklisting_start_time = total_max_time * blacklist_trigger_percentage;
+    while (true) {
+        // Check if we need to start blacklisting.
+        if (!blacklisting && timer.get_elapsed_time() > blacklisting_start_time) {
+            blacklisting = true;
+            // Also reset stagnation timer in case it was already set.
+            stagnation_start_time = -1;
+            if (verbosity >= utils::Verbosity::NORMAL) {
+                utils::g_log << "given percentage of total time limit "
+                             << "exhausted; enabling blacklisting." << endl;
+            }
+        }
+
+        unordered_set<int> blacklisted_variables;
+        if (blacklisting && !non_goal_variables.empty()) {
+            /*
+              Randomize the number of non-goal variables for blacklisting.
+              We want to choose at least 1 non-goal variable and up to the
+              entire set of non-goal variables.
+            */
+            int blacklist_size = (*rng)(non_goal_variables.size());
+            ++blacklist_size; // [1, |non-goal variables|]
+            rng->shuffle(non_goal_variables);
+            for (int i = 0; i < blacklist_size; ++i) {
+                int var_id = non_goal_variables[i];
+                blacklisted_variables.insert(var_id);
+            }
+            if (verbosity >= utils::Verbosity::DEBUG) {
+                utils::g_log << "blacklisting " << blacklist_size << " out of "
+                             << non_goal_variables.size()
+                             << " non-goal variables: ";
+                for (int var : blacklisted_variables) {
+                    utils::g_log << var << ", ";
+                }
+                utils::g_log << endl;
+            }
+        }
+
+        int remaining_pdb_size_for_cegar = min(remaining_collection_size, max_pdb_size);
+        double remaining_time_for_cegar = min(timer.get_remaining_time().operator double(), cegar_max_time);
```

Comment:
Fancy, I've never seen something like this before ;)

Reply:
It's probably more idiomatic to write `static_cast<double>(timer.get_remaining_time())`

Reply:
I agree. I actually don't recall writing this myself; I assume my IDE did when I was refactoring something.